OSOE-49: Replace Windows-only code in UI Testing Toolbox, upgrade to latest Atata/Selenium

sarahelsaig opened the pull request:
[OSOE-49](https://lombiq.atlassian.net/browse/OSOE-49)

Fixes #123.

## Release Notes

- Please add these new lines to your _.gitattributes_, because we get IDE0055 warnings when the C# source file has Unix style line endings. Some .cs files will be affected, please run `git add --renormalize .` on your projects and `git submodule foreach git add --renormalize .` if they include submodules.
- The setup and configuration for using SQL Server from a Docker container has changed. See new instructions [here](https://github.com/Lombiq/UI-Testing-Toolbox/blob/e764b5c3cb4e40d3a9043b520e71d0c5c44173f7/Lombiq.Tests.UI/Docs/Configuration.md#using-sql-server-from-a-docker-container). You can also find a sample config and setup steps in the _.github_ directory.
- We have upgraded to Atata 2 which in turn upgraded to Selenium 4. Both introduce breaking changes, although most of it is pretty easy to fix. Replace every `RemoteWebDriver` with `IWebDriver` and if you execute scripts or take screenshots then use the extension methods ([here](https://github.com/Lombiq/UI-Testing-Toolbox/blob/e764b5c3cb4e40d3a9043b520e71d0c5c44173f7/Lombiq.Tests.UI/Extensions/ScreenshotUITestContextExtensions.cs), [here](https://github.com/Lombiq/UI-Testing-Toolbox/blob/e764b5c3cb4e40d3a9043b520e71d0c5c44173f7/Lombiq.Tests.UI/Extensions/ScriptingUITestContextExtensions.cs) and [here](https://github.com/Lombiq/UI-Testing-Toolbox/blob/e764b5c3cb4e40d3a9043b520e71d0c5c44173f7/Lombiq.Tests.UI/Extensions/ScriptingWebDriverExtensions.cs)).
- If you have several UI test projects it can be cumbersome to maintain a _TestConfiguration.json_ file for each. Instead, you can set the value of the `LOMBIQ_UI_TESTING_TOOLBOX_SHARED_TEST_CONFIGURATION` environment variable to the absolute path of a central configuration file and then each project will look it up. If you place an individual _TestConfiguration.json_ into a test directory it will still take precedence in case you need special configuration for just that one.
- New _HelpfulLibraries.Cli_ project has been added, please use and expand it for all your `Process` and `CliWrap` needs.

sarahelsaig committed:
Adding explicit "default" Langversion to project file.

sarahelsaig committed:
Fix windows-specific executable invocation and replace exe terminology.

sarahelsaig committed:
Avoid unintended reference comparison.

sarahelsaig committed:
Add WithArgumentsAsync extension and some code cleanup.

sarahelsaig committed:
Get version info using CLI instead of the built-in Registry based solution.

sarahelsaig committed:
More helpful exception message for when NPM lets out the magic smoke.

sarahelsaig committed:
Ensures the started process runs in the Development environment as many features and shortcuts rely on that.

sarahelsaig committed:
Fix .exe usage in SmtpService.

sarahelsaig committed:
Add OperatingSystemHelpers to DRY code.

sarahelsaig committed:
Eliminate unnecessarily Windowsy terminology from comments.

sarahelsaig committed:
Linux specific documentation.

sarahelsaig committed:
Fix some SQL server errors.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Post merge fixup.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Fix gitattributes.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
More code styling.

sarahelsaig committed:
Normalize all the line endings

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Add NVM related docs.

sarahelsaig committed:
Fix webroot argument that somehow corrupted the following environment argument.

sarahelsaig committed:
Disable UseSqlServer in BasicTests.

There is no point because there are separate SQL Server specific tests.

sarahelsaig committed:
Add linux notes about SQL Server.

sarahelsaig committed:
Add LOMBIQ_SHARED_TEST_CONFIGURATION and doc it.

sarahelsaig committed:
Fix host path problem with tilde.

sarahelsaig committed:
Fix file access problem during SQL backup.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Merge remote-tracking branch 'origin/issue/OSOE-60' into issue/OSOE-49

# Conflicts:
#	Lombiq.Tests.UI.Samples/Tests/AzureBlobStorageTests.cs
#	Lombiq.Tests.UI.Samples/Tests/BasicTests.cs
#	Lombiq.Tests.UI.Samples/Tests/SqlServerTests.cs
#	Lombiq.Tests.UI/Extensions/CommandExtensions.cs
#	Lombiq.Tests.UI/Extensions/HtmlValidationUITestContextExtensions.cs
#	Lombiq.Tests.UI/Lombiq.Tests.UI.csproj
#	Lombiq.Tests.UI/Services/AtataFactory.cs
#	Lombiq.Tests.UI/Services/IWebApplicationInstance.cs
#	Lombiq.Tests.UI/Services/OrchardCoreInstance.cs
#	Lombiq.Tests.UI/Services/SmtpService.cs
#	Lombiq.Tests.UI/Services/SqlServerManager.cs
#	Lombiq.Tests.UI/Services/TestConfigurationManager.cs
#	Lombiq.Tests.UI/Services/UITestExecutionSession.cs
#	Lombiq.Tests.UI/Services/WebDriverFactory.cs

sarahelsaig committed:
Post merge fixup.

sarahelsaig committed:
Fix TargetFramework.

sarahelsaig committed:
Add a new line at the end of the file

sarahelsaig committed:
Merge branch 'dev' into issue/OSOE-49

# Conflicts:
#	Lombiq.Tests.UI.Samples/Tests/BasicTests.cs
#	Lombiq.Tests.UI/Docs/Configuration.md
#	Lombiq.Tests.UI/Services/OrchardCoreInstance.cs
#	Lombiq.Tests.UI/Services/UITestExecutionSession.cs

sarahelsaig committed:
Fix namespace.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Fix Linux part of Configuration.md.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Enforce Windows newlines for C# files to avoid false positives with IDE55 warning.

sarahelsaig committed:
Fix missing inner exception for fast fail setup.

sarahelsaig committed:
Elaborate on CreateDriverAsync failure.

sarahelsaig committed:
Make the Chrome driver version picking more forgiving.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Use version api uri.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Refactor GetVersion to actually work correctly on Windows.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

0liver commented on Lombiq.Tests.UI/Services/WebDriverFactory.cs:
```diff
@@ -167,4 +176,60 @@ private class StaticVersionEdgeConfig : EdgeConfig
+        static (string Executable, string[] Arguments) GetWithCli(string name) =>
+            (name, new[] { "--version" });
+
+        var isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+        var (executableName, arguments) = (driverConfig, isWindows) switch
+        {
+            (ChromeConfig, true) => GetWithPowershell("chrome"),
+            (ChromeConfig, false) => GetWithCli("chromium"),
+            (FirefoxConfig, true) => GetWithPowershell("firefox"),
+            (FirefoxConfig, false) => GetWithCli("firefox"),
+            (EdgeConfig, true) => ("powershell", new[] { "(Get-AppxPackage Microsoft.MicrosoftEdge).Version" }),
+            (InternetExplorerConfig, _) =>
+                ("reg", new[] { "query", @"HKEY_LOCAL_MACHINE\Software\Microsoft\Internet Explorer", "/v", "svcVersion" }),
+            _ => throw new ArgumentOutOfRangeException(nameof(driverConfig), driverConfig, message: null),
+        };
+
+        string version;
+
+        try
+        {
+            executableName = (await CliHelper.WhichAsync(executableName)).First().FullName;
+
+            var result = await Cli
+                .Wrap(executableName)
+                .WithArguments(arguments)
+                .ExecuteBufferedAsync();
+            version = result.StandardOutput.Split().Reverse().FirstOrDefault(word => Version.TryParse(word, out _));
+
+            if (version != null && driverConfig is ChromeConfig && Version.TryParse(version, out var chromeVersion))
+            {
+                // Chrome doesn't always have the driver for the same revision you have installed, especially if you use
+                // a package manager instead of the browser's auto-updater. To get around this, interrogate the API for
+                // the latest revision of the same build version, as that should be still compatible.
+                var versionApiUri = new Uri(
+                    "https://chromedriver.storage.googleapis.com/LATEST_RELEASE_" +
+                    FormattableString.Invariant($"{chromeVersion.Major}.{chromeVersion.Minor}.{chromeVersion.Build}"));
+                using var client = new HttpClient();
+                version = await (await client.GetAsync(versionApiUri)).Content.ReadAsStringAsync();
+            }
+        }
```
👁️ WoW 👁️

sarahelsaig committed:
Fix comment in gitattributes.

sarahelsaig committed:
Whitelist IPv6 localhost.

sarahelsaig committed:
Resolve localhosts.

sarahelsaig committed:
Add disable-dev-shm-usage switch.

sarahelsaig committed:
Don't download chrome driver if it's already installed.

sarahelsaig committed:
Force headless mode if we are in Linux without a working graphical environment.

sarahelsaig committed:
Remove debug output.

sarahelsaig committed:
Fix truncated smtp4dev error reporting.

sarahelsaig committed:
Unusing.

sarahelsaig committed:
Improve Docker SQL Server configuration readme.

sarahelsaig committed:
Use DotnetCli.

sarahelsaig committed:
Probe LOMBIQ_SHARED_TEST_CONFIGURATION.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Restore missing HostSnapshotPath fix.

When did this vanish wtf?!

sarahelsaig committed:
Turn AssertLogs into extension and call it before verifying setup success.

sarahelsaig committed:
Note on PlatformNotSupportedException.

sarahelsaig committed:
Rewrite docker db handling to not use bind.

sarahelsaig committed:
unusing

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Unusing

sarahelsaig committed:
Format JSON in Configuration.md.

sarahelsaig committed:
Reset ownership for the backup copied into docker.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Bug fix.

sarahelsaig committed:
Bug fix.

sarahelsaig committed:
More bug fix, why can't I type?!

sarahelsaig committed:
Upgrade Atata to 2.0.0

sarahelsaig committed:
Code cleanup after Atata upgrade.

sarahelsaig committed:
Bug fix and refactor LoggingWebDriverExtensions.GetEndpoint

sarahelsaig committed:
Add default ProgramCli configuration and document it.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Upgrade packages.

sarahelsaig committed:
Use UITestContext.ExecuteScript everywhere.

sarahelsaig committed:
Fix context.TakeScreenshot().

sarahelsaig committed:
Fix CreateEdgeDriverAsync after Selenium upgrade.

sarahelsaig committed:
Obsolete GetAndEmptyBrowserLogAsync, replace with GetAndEmptyBrowserLog

sarahelsaig committed:
Temporarily disable emailing test.

sarahelsaig committed:
Fix monkey testing after the log update.

sarahelsaig committed:
Fix test.

sarahelsaig committed:
Prevent lockup.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Upgrade rnwood.smtp4dev and re-enable test.

sarahelsaig committed:
Exclude only GitHub Actions from the email test.

sarahelsaig committed:
update gitattributes

sarahelsaig committed:
Documentation improvements.

Piedone commented on .gitattributes:
```diff
@@ -1 +1,5 @@
 * text=auto
+
+# Enforce Windows newlines for C# files to avoid false positives with IDE0055 warning.
+# See https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/issues/106
+*.cs text eol=crlf
```
Also add this to https://github.com/Lombiq/Open-source-template/blob/dev/.gitattributes

sarahelsaig replied:
[Done](https://github.com/Lombiq/Open-source-template/pull/2)

Piedone commented on Lombiq.Tests.UI.Samples/Helpers/SetupHelpers.cs:
```diff
@@ -61,9 +61,21 @@ public static async Task RunSetupAndSignInDirectlyAsync(UITestContext context, s
+        catch
+        {
+            // Before throwing a not-too-helpful NoSuchElementException, verify if the browser log has something to say.
+            await context.AssertLogsAsync();
```
This should already happen due to the `AfterPageChange` event being raised. If that's not fired, then do it with `TriggerAfterPageChangeEventAsync()` after setup. The best would be some generic way, so it's triggered after every setup, but I'm not sure we can say that the result of `SetupOperation` will always be something that can be considered a "page". But since we require it to return a URL anyway, seems fair I think.

sarahelsaig replied:
On further investigation I was wrong about this. Actually the problem is, that the setup doesn't log into file, just the console. Something is not right here. The OrchardCore.Cms.Web 1.3.0 doesn't behave like this so it's something on our end...

Piedone replied:
I believe this used before (and tests failed due to exceptions during setup, even if nothing got broken visibly) so please investigate.

sarahelsaig replied:
I have just verified that this problem is present on the current dev branch as well. See demo of it happening on both normal execution and UI test below. So this is not related to the current issue, and I'd like to handle it in a separate issue to not hold back this one any more. Are you okay with that?

https://user-images.githubusercontent.com/4007293/168492718-36f9b391-e511-470c-8bf3-6e88d7a80f77.mp4

https://user-images.githubusercontent.com/4007293/168492720-4db5d864-91b3-4843-996f-e283afda4457.mp4

As a short term fix [I have made it look for validation errors](https://github.com/Lombiq/UI-Testing-Toolbox/pull/119/commits/9bcd690d5502575ab915ba65d657d225edb09221). This would've been helpful for the problem I had with failed DB setup during development and looks a bit more intelligible:
![image](https://user-images.githubusercontent.com/4007293/168492825-da9449d9-8ec1-4978-bc45-24d911646027.png)

Piedone replied:
I see. This is not what I meant but rather, exceptions when the setup is already running (like module migrations failing) I remember being logged and failing tests before. I can't recall that we ever checked what happens in the scenario you're showing.

OK, but please open an OSOE issue and start working on this afterwards.

Piedone replied:
The fix and new issue is good, I'm just leaving this convo unresolved so it can be linked from the new issue.

Piedone commented on Lombiq.Tests.UI.Samples/Tests/AzureBlobStorageTests.cs:
```diff
@@ -1,17 +1,16 @@
 
-// Up until now the Orchard app always used the default local Media storage for managing Media files. However, you may
-// use Azure Blob Storage in production. You can also test your app with it!
+// Up until now the Orchard app always used the default local Media storage for managing Media files. However, you
+// may use Azure Blob Storage in production. You can also test your app with it!
```
The original was good.

sarahelsaig replied:
This file was already edited before we went over to file scoped namepsaces, so I had to run the comment realign from Code Maid separately. Guess it didn't work out so well, becuase I have already fixed a bunch of these manually yet you still found more :(

Piedone replied:
Many, many more :(. big sad

sarahelsaig replied:
Where? I can't see any. Besides the two I have intentionally left and commented on below.

Piedone replied:
Was referring to the ones I commented and fixed.

Piedone commented on Lombiq.Tests.UI.Samples/Tests/AzureBlobStorageTests.cs:
```diff
@@ -20,20 +19,17 @@ public AzureBlobStorageTests(ITestOutputHelper testOutputHelper)
-    // site's Media storage. If they still work and there are no errors in the log then the app works with Azure Blob
-    // Storage too.
+    // site's Media storage. If they still work and there are no errors in the log then the app works with Azure
+    // Blob Storage too.
```
The original was good.

Piedone commented on Lombiq.Tests.UI.Samples/Tests/AzureBlobStorageTests.cs:
```diff
@@ -20,20 +19,17 @@ public AzureBlobStorageTests(ITestOutputHelper testOutputHelper)
+            // Note the configuration! We could also set this globally in UITestBase.
+            // You'll need an accessible Azure Blob Storage account. For testing we recommend the Azurite emulator
+            // (https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azurite) that can be used from
+            // tests without any further configuration.
```
The original was good.

Piedone commented on Lombiq.Tests.UI.Samples/Tests/EmailTests.cs:
```diff
@@ -20,7 +22,12 @@ public EmailTests(ITestOutputHelper testOutputHelper)
+        // Currently smtp4dev throws an exception on GitHub Actions. The same happens even if you call it directly, it's
+        // not related to this app. See issue: https://github.com/rnwood/smtp4dev/issues/1064
+        RuntimeInformation.IsOSPlatform(OSPlatform.Linux) &&
+        !string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("GITHUB_ENV"))
```
```suggestion
        RuntimeInformation.IsOSPlatform(OSPlatform.Linux) &&
            !string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("GITHUB_ENV"))
```

Piedone commented on Lombiq.Tests.UI.Samples/Tests/EmailTests.cs:
```diff
@@ -20,7 +22,12 @@ public EmailTests(ITestOutputHelper testOutputHelper)
     public Task SendingTestEmailShouldWork(Browser browser) =>
-        ExecuteTestAfterSetupAsync(
+        // Currently smtp4dev throws an exception on GitHub Actions. The same happens even if you call it directly, it's
+        // not related to this app. See issue: https://github.com/rnwood/smtp4dev/issues/1064
```
This is supposed to be fixed now.

sarahelsaig replied:
I think you were viewing a previous commit becuase it was fixed more or less the same time you reviewed,

Piedone replied:
Yep.

sarahelsaig committed:
Upgrade to Rnwood.Smtp4dev 3.1.4

Piedone commented on Lombiq.Tests.UI.Samples/Tests/ErrorHandlingTests.cs:
```diff
@@ -64,6 +66,40 @@ public ErrorHandlingTests(ITestOutputHelper testOutputHelper)
             },
             browser);
+
+    // To be able to trust the test above, we have to be sure that the browser logs survive the navigation events and
+    // all get collected into the historic browser log.
```
When is this not working?

sarahelsaig replied:
It's to prove that the new/fixed browser logging from Selenium 4 (`driver.Manage().Logs.GetLog(LogType.Browser);`) and my extension new method using it actually works.

Piedone commented on Lombiq.Tests.UI.Samples/Tests/SqlServerTests.cs:
```diff
@@ -11,10 +10,10 @@
+// tests will get their DBs there (if you run your app with SQL Server in production then it's recommended to also
+// test with it, should there be any incompatibilities). Note that for this, you need an SQL Server instance
+// running; by default, this will be attempted under the default localhost server name. If you're using anything
+// else, check out the settings in SqlServerConfiguration and Docs/Configuration.md, especially if you use Docker.
```
Original was correct.

Piedone commented on Lombiq.Tests.UI/Docs/Configuration.md:
```diff
@@ -49,15 +49,17 @@ Here's a full *TestConfiguration.json* file example, something appropriate durin
 
 We encourage you to experiment with a `RetryTimeoutSeconds` value suitable for your hardware. Higher, paradoxically, is usually less safe.
 
+If you have several UI test projects it can be cumbersome to maintain a *TestConfiguration.json* file for each. Instead you can set the value of the `LOMBIQ_SHARED_TEST_CONFIGURATION` environment variable to the absolute path of a central configuration file and then each project will look it up. If you place an individual *TestConfiguration.json* into a test directory it will still take precedenc ...
```
Nice!

Piedone commented on Lombiq.Tests.UI/Services/TestConfigurationManager.cs:
```diff
@@ -62,9 +62,17 @@ public static T GetConfiguration<T>()
+        return new ConfigurationBuilder()
             .AddJsonFile("TestConfiguration.json", optional: true, reloadOnChange: false)
+            .AddJsonFile(
+                Environment.GetEnvironmentVariable("LOMBIQ_SHARED_TEST_CONFIGURATION") ?? $"dummy_{Guid.NewGuid()}",
```
Let's call it `LOMBIQ_UI_TESTING_TOOLBOX_SHARED_TEST_CONFIGURATION` instead.

Piedone commented on Lombiq.Tests.UI/Services/TestConfigurationManager.cs:
```diff
@@ -62,9 +62,17 @@ public static T GetConfiguration<T>()
-        new ConfigurationBuilder()
+    private static IConfiguration BuildConfiguration()
+    {
+        Console.WriteLine(
```
Seems like a leftover debug code.

Piedone commented on Lombiq.Tests.UI/Docs/Configuration.md:
```diff
@@ -73,27 +75,49 @@ If you have multiple UI test projects in a single solution and you're executing
+
+#### On Both
+
+If you want to test it out, type `docker exec -u 0 -it sql2019 /opt/mssql-tools/bin/sqlcmd -S localhost -U sa -P "yourStrong(!)Password"` to access the SQL console.
```
How about using "Password1!" for this too?

Piedone commented on Lombiq.Tests.UI/Extensions/CommandExtensions.cs:
```diff
@@ -8,8 +9,8 @@ namespace CliWrap;
-    /// Executes a <see cref="Command"/> as a dotnet.exe command that starts a long-running application, and waits for
-    /// the app to be started.
+    /// Executes a <see cref="Command"/> as a <c>dotnet</c> command that starts a long-running application, and
+    /// waits for the app to be started.
```
Line too short. You have something configured for too short auto-formatting, so please go through all changed comments.

Piedone commented on Lombiq.Tests.UI/Extensions/CommandExtensions.cs:
```diff
@@ -32,4 +33,16 @@ public static class CommandExtensions
+    /// Same as <see cref="Command.WithArguments(Action{ArgumentsBuilder})"/>, but the configuration delegate is
+    /// asynchronous.
+    /// </summary>
+    public static async Task<Command> WithArgumentsAsync(this Command command, Func<ArgumentsBuilder, Task> configureAsync)
```
Wouldn't this and the whole class be better in `Lombiq.HelpfulLibraries.Cli`?

Piedone commented on Lombiq.Tests.UI/Extensions/ScriptingUITestContextExtensions.cs:
```diff
@@ -1,16 +1,16 @@
 
     public static object ExecuteAsyncScript(this UITestContext context, string script, params object[] args) =>
-        context.ExecuteLogged(nameof(ExecuteAsyncScript), script, () => ((IJavaScriptExecutor)context.Driver).ExecuteAsyncScript(script, args));
+        context.ExecuteLogged(nameof(ExecuteAsyncScript), script, () => context.Driver.AsScriptExecutor().ExecuteAsyncScript(script, args));
```
I now see that we have several of these `AsScriptExecutor()`s so an extension would be nice still.

DAud-IcI replied:
We only have 3 and out of that 2 are in this file for the context extensions, so I don't think we need it.

Piedone commented on Lombiq.Tests.UI/Lombiq.Tests.UI.csproj:
```diff
@@ -73,6 +72,7 @@
   </ItemGroup>
 
   <ItemGroup Condition="'$(NuGetBuild)' != 'true'">
+    <ProjectReference Include="..\..\..\src\Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.Cli\Lombiq.HelpfulLibraries.Cli.csproj" />
```
Also add to the NuGet references.

Piedone commented on Lombiq.Tests.UI/MonkeyTesting/MonkeyTester.cs:
```diff
@@ -168,24 +167,24 @@ private TimeSpan TestCurrentPageAndMeasureTestTimeLeft(TimeSpan testTime, int ra
         var testTimeLeft = MeasureTimeLeftOfMeetingPredicate(
             _context.Driver,
-            driver => !(bool)driver.ExecuteScript(GremlinsScripts.GetAreGremlinsRunningScript),
+            _ => !(bool)_context.ExecuteScript(GremlinsScripts.GetAreGremlinsRunningScript),
```
The driver might be different within `SafeWait` so safer to use that.

Piedone commented on Lombiq.Tests.UI/OrchardCoreUITestBase.cs:
```diff
@@ -261,6 +262,10 @@ public abstract class OrchardCoreUITestBase
             BrowserConfiguration = { Browser = browser },
         };
 
+        ProgramCli.DefaultShellCliCommandFactory = OSDependentShellCliCommandFactory
+            .UseCmdForWindows()
+            .UseForOtherOS(new BashShellCliCommandFactory("-login"));
```
Does this need to be configured statically? If yes, why not in the static ctor? Also, it would be better not here but in `AtataFactory`.

sarahelsaig replied:
Yes, it's a static config. I moved it to the static ctor.

Piedone commented on Lombiq.Tests.UI/Services/OrchardCoreInstance.cs:
```diff
@@ -238,14 +245,18 @@ private async Task StartOrchardAppAsync()
+                var note = stdErr.Text.ContainsOrdinalIgnoreCase("Failed to bind to address")
+                    ? " This can happen when there are leftover " + dotnet +
+                      " processes after an aborted test run or some other app is listening on the same port too."
+                    : string.Empty;
```
Some note about what might be an issue or where to look?

sarahelsaig replied:
Look, I just moved the text out of the exception constructor to make it more readable. [Ask the person who wrote it originally](https://github.com/Lombiq/UI-Testing-Toolbox/commit/5acd4781f8484cc99d631f515dc96c620ca97738), I hear you are on good terms. 😁

Piedone replied:
To paraphrase Knuth, beware of bugs in the above code; I have only written it, not reviewed it. :D

Piedone commented on Lombiq.Tests.UI/Services/SqlServerManager.cs:
```diff
@@ -16,13 +20,13 @@ public class SqlServerConfiguration
         "SqlServerDatabaseConfiguration:ConnectionStringTemplate",
         $"Server=.;Database=LombiqUITestingToolbox_{DatabaseIdPlaceholder};Integrated Security=True;" +
-            $"MultipleActiveResultSets=True;Connection Timeout=60;ConnectRetryCount=15;ConnectRetryInterval=5");
+        $"MultipleActiveResultSets=True;Connection Timeout=60;ConnectRetryCount=15;ConnectRetryInterval=5");
```
This was intentional, so it's a different level than the parameters.

sarahelsaig replied:
It's not standard though.

Piedone replied:
We frequently do such indentations.

sarahelsaig replied:
I guess not frequently enough cause I've never seen 'em.

Piedone replied:
This is typical with `if`s with complex clauses, like (obviously if the lines are long):

```c#
if (bla &&
    (bla ||
        bla) &&
    bla)
{
```

sarahelsaig replied:
Yes, I have seen that kind, but these are different things. In your example there are parentheses so indentation is logical.

Piedone replied:
And here we have the multiple parameters. The first line is a parameter, the second line is a parameter, the third line is... Part of the second parameter. You get the idea.

sarahelsaig replied:
![image](https://user-images.githubusercontent.com/4007293/168440575-6eaa038d-78ed-471e-aaf3-5b3299e55d8a.png)

Piedone replied:
:D

Piedone commented on Lombiq.Tests.UI/Services/SqlServerManager.cs:
```diff
@@ -85,39 +90,46 @@ public SqlServerRunningContext CreateDatabase()
+            Path.GetDirectoryName(filePathLocal) ??
+            throw new InvalidOperandException($"Failed to get the directory path for local path \"{filePathLocal}\".");
 
+        if (!Directory.Exists(directoryPathLocal)) Directory.CreateDirectory(directoryPathLocal);
```
Use `FileSystemHelper.EnsureDirectoryExists()` instead.

Piedone commented on Lombiq.Tests.UI/Services/SqlServerManager.cs:
```diff
@@ -85,39 +90,46 @@ public SqlServerRunningContext CreateDatabase()
 
         var useCompression = useCompressionIfAvailable &&
-            (server.EngineEdition == Edition.EnterpriseOrDeveloper || server.EngineEdition == Edition.Standard);
+                             (server.EngineEdition == Edition.EnterpriseOrDeveloper || server.EngineEdition == Edition.Standard);
```
That's a looong indentation.

sarahelsaig replied:
It wants to indent it to the same column where the previous expression starts (so after the `=`). That's why you have to break the first part into a new line as well:
```csharp
        var useCompression =
            useCompressionIfAvailable &&
            (server.EngineEdition == Edition.EnterpriseOrDeveloper || server.EngineEdition == Edition.Standard);
```

Piedone replied:
Yeah but adjusting to the column like this is something we never do (for a reason) though I know you like it ;). I'm OK with the second version too.

Piedone commented on Lombiq.Tests.UI/Services/SqlServerManager.cs:
```diff
@@ -139,9 +151,18 @@ public SqlServerRunningContext CreateDatabase()
-        backup.SqlBackup(server);
+        // We could use SqlBackupAsync() too but that's not Task-based async, we'd need to subscribe to an event
+        // which is messy.
+        await Task.Run(() => backup.SqlBackup(server));
```
Why is this sync-in-async necessary?

sarahelsaig replied:
Because otherwise AsyncFixer will complain why we don't use `backup.SqlBackupAsync` instead. Also this is a long sync call so why there is no reason not to run it in a separate thread.

Piedone replied:
A suppression here would be better then.

I don't see why running this on a separate thread (if at all) would bring any benefits. We're blocking one thread in any case, and the rest of the code can't continue anyway.

Piedone commented on Lombiq.Tests.UI/Services/SqlServerManager.cs:
```diff
@@ -153,7 +174,10 @@ public SqlServerRunningContext CreateDatabase()
     }
 
-    public void RestoreSnapshot(string snapshotDirectoryPath)
+    public async Task RestoreSnapshotAsync(
```
Try a failing SQL Server test to see if failure dump creation works properly too.

sarahelsaig replied:
Works, here: https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/actions/runs/2325275960

Piedone commented on Lombiq.Tests.UI/Services/UITestExecutionSession.cs:
```diff
@@ -130,7 +128,7 @@ public async Task<bool> ExecuteAsync(int retryCount, string dumpRootPath)
                 _testOutputHelper.WriteLineTimestampedAndDebug(
                     "The test was attempted {0} time(s) and won't be retried anymore. You can see more details " +
-                        "on why it's failing in the FailureDumps folder: {1}",
+                    "on why it's failing in the FailureDumps folder: {1}",
```
The indentation was a distinction from the parameters.

Piedone commented on Lombiq.Tests.UI/Services/UITestExecutionSession.cs:
```diff
@@ -392,14 +391,14 @@ private async Task SetupAsync()
                 if (setupConfiguration.FastFailSetup &&
-                    _setupOperationFailureCount.TryGetValue(GetSetupHashCode(), out var failureCount) &&
-                    failureCount > _configuration.MaxRetryCount)
+                    _setupOperationFailureCount.TryGetValue(GetSetupHashCode(), out var pair) &&
```
```suggestion
                    _setupOperationFailureCount.TryGetValue(GetSetupHashCode(), out var failure) &&
```

Piedone commented on Lombiq.Tests.UI/Services/UITestExecutionSession.cs:
```diff
@@ -392,14 +391,14 @@ private async Task SetupAsync()
+                    pair.FailureCount > _configuration.MaxRetryCount)
                 {
-                    throw new SetupFailedFastException(failureCount);
+                    throw new SetupFailedFastException(pair.FailureCount, pair.LatestException);
```
Very useful addition!

Piedone commented on Lombiq.Tests.UI/Services/WebDriverFactory.cs:
```diff
@@ -18,34 +25,60 @@ public static class WebDriverFactory
-                var options = new ChromeOptions().SetCommonOptions();
+    public static Task<ChromeDriver> CreateChromeDriverAsync(BrowserConfiguration configuration, TimeSpan pageLoadTimeout)
+    {
+        // Force headless mode if we are in Linux without a working graphical environment.
+        if (!configuration.Headless && RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+        {
+            configuration.Headless = string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("DISPLAY"));
+        }
```
This looks like something for all drivers.

sarahelsaig replied:
Good point, I have moved it to `AtataFactory.CreateDriverAsync` before `WebDriverFactory.CreateChromeDriverAsync` and friends are called.

Piedone commented on Lombiq.Tests.UI/Services/WebDriverFactory.cs:
```diff
@@ -116,10 +142,10 @@ private static TDriverOptions SetCommonOptions<TDriverOptions>(this TDriverOptio
     {
         // Setting timeouts for cases when tests randomly hang up a bit more for some reason (like the test machine load
-        // momentarily spiking). We're not increasing ImplicityWait, the default of which is 0, since that would make
+        // momentarily spiking). We're not increasing ImplicitlyWait, the default of which is 0, since that would make
```
Convention-over-configuration enthusiasts live in the Implicity.

Piedone commented on Lombiq.Tests.UI/Services/WebDriverFactory.cs:
```diff
@@ -128,37 +154,37 @@ private static TDriver SetCommonTimeouts<TDriver>(this TDriver driver, TimeSpan
             // https://stackoverflow.com/a/31637510/220230
             _ = _driverSetups.GetOrAdd(driverConfig.GetName(), _ => new Lazy<bool>(() =>
             {
-                // Version selection based on the locally installed version is only available for Chrome, see:
-                // https://github.com/rosolko/WebDriverManager.Net/pull/91.
-                if (driverConfig is ChromeConfig)
-                {
-                    new DriverManager().SetUpDriver(driverConfig, VersionResolveStrategy.MatchingBrowser);
-                }
-                else
-                {
-                    new DriverManager().SetUpDriver(driverConfig);
-                }
```
Why is this not necessary anymore?

sarahelsaig replied:
The version selection has been moved into the TryFindVersionAsync method so we always have a version parameter.

Piedone commented on Lombiq.Tests.UI/Services/WebDriverFactory.cs:
```diff
@@ -167,4 +193,66 @@ private class StaticVersionEdgeConfig : EdgeConfig
         public override string GetLatestVersion() => "83.0.478.37";
     }
+
+    private static async Task<string> TryFindVersionAsync(IDriverConfig driverConfig)
```
The Windows logic for this is included in WebDriverManager.

sarahelsaig replied:
What part?

Piedone replied:
WebDriverManager auto-selects the matching version for Chrome on Windows.

sarahelsaig replied:
Yeah, but it uses Regex which I don't trust. If chrome is in the path, my solution finds it with powershell, and if it's not it will still fall back to the MatchingBrowser thing from WebDriverManager

Piedone replied:
Alright :D.

![image](https://user-images.githubusercontent.com/1976647/168443427-e52163d8-be70-44e5-b278-16960231746d.png)

sarahelsaig replied:
lol typo, meant to write regedit. I trust regex implicitly :D

sarahelsaig replied:
I did some testing and now I understand why you actually need regedit there. Apparently Windows now logs app paths in the registry even for apps that aren't in the `%PATH%`. How? No idea.

Piedone commented on Lombiq.Tests.UI/Services/WebDriverFactory.cs:
```diff
@@ -167,4 +193,66 @@ private class StaticVersionEdgeConfig : EdgeConfig
+
+    private static async Task<string> TryFindVersionAsync(IDriverConfig driverConfig)
+    {
+        static (string Executable, string[] Arguments) GetWithPowershell(string name) =>
```
All this looks very useful for WebDriverManager too. Please add an issue there about adding support for more browsers, referencing this code.

0liver replied:
Could you link that issue here, please?

sarahelsaig replied:
https://github.com/rosolko/WebDriverManager.Net/pull/186

sarahelsaig replied:
Good news, my WebDriverManager.Net PR just got approved by the owner.
Not merged yet for some reason, but we'd have to wait until the next release to make use of it anyway so I don't mind that.

0liver replied:
It's merged now 😉

sarahelsaig replied:
It has been released too. I will update our deps soon.

Piedone committed:
Code styling

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Strip invalid path characters to please actions/upload-artifact@v2.

sarahelsaig committed:
Revert "Strip invalid path characters to please actions/upload-artifact@v2."

This reverts commit 8f17bd0c2b1d85f690ade2aaeece58c96444490c.

sarahelsaig committed:
Improve MakeFileSystemFriendly.

sarahelsaig committed:
Add missing badded character.

sarahelsaig committed:
Comment fixes.

sarahelsaig committed:
Rename shared config path environment variable.

sarahelsaig commented on Lombiq.Tests.UI/Services/SmtpService.cs:
```diff
@@ -61,8 +60,8 @@ public async Task<SmtpServiceRunningContext> StartAsync()
-        // The local tool manifest was already created with dotnet new tool-manifest and the tool installed with: dotnet
-        // tool install Rnwood.Smtp4dev --version "3.1.0-*"
+        // The local tool manifest was already created with dotnet new tool-manifest and the tool installed with:
+        // dotnet tool install Rnwood.Smtp4dev --version "3.1.0-*"
```
I won't change this back becuase it's more readable. Same for
https://github.com/Lombiq/UI-Testing-Toolbox/pull/119/files#diff-0b2964ca5d70a7ff5a4c7473f25ab93e6ecea50a60c67568c20d21b570e68360R108

sarahelsaig committed:
Format comments.

sarahelsaig committed:
Clean up TakeSnapshotAsync's documentation.

sarahelsaig committed:
Fix more comments.

sarahelsaig committed:
Move CommandExtensions.

sarahelsaig committed:
Add and use ScriptingWebDriverExtensions

sarahelsaig committed:
Code styling and various small improvements.

sarahelsaig committed:
No Task Run.

sarahelsaig committed:
How about using "Password1!" for this too?

sarahelsaig committed:
Hopefully fix non-Docker SQL.

sarahelsaig committed:
Intentional test error in SQL test.

sarahelsaig committed:
Revert

sarahelsaig committed:
Rename CommandAsync to ExecuteAsync

sarahelsaig committed:
Add Cli NuGet package.

sarahelsaig committed:
Display setup validation errors.

Piedone commented on Lombiq.Tests.UI.Samples/Helpers/SetupHelpers.cs:
```diff
@@ -61,21 +64,31 @@ public static async Task RunSetupAndSignInDirectlyAsync(UITestContext context, s
+                    .Select(element => "\n- " + element.Text.Trim())
+                    .Join(string.Empty);
+
+            throw new InvalidOperationException($"Setup has failed with the following validation error{plural}:{errors}");
```
An assertion exception or something would be more suitable.

sarahelsaig committed:
Use AssertionException.

sarahelsaig committed:
Unusing

sarahelsaig committed:
Simplify TryFindVersionAsync.

sarahelsaig committed:
Cleanup.

sarahelsaig committed:
Code cleanup

sarahelsaig committed:
Add comment about WebDriverManager.Net PR.

Piedone commented on Lombiq.Tests.UI/Lombiq.Tests.UI.csproj:
```diff
@@ -51,20 +51,19 @@
-    <PackageReference Include="CliWrap" Version="3.4.1" />
+    <PackageReference Include="Atata" Version="2.0.0" />
+    <PackageReference Include="Atata.Bootstrap" Version="2.0.0" />
+    <PackageReference Include="Atata.HtmlValidation" Version="2.0.0" />
```
As I learned from Jocó, the config format of `html-validate` changed for 7.x. The version behind the scenes changed for `Atata.HtmlValidation` also changed from 6.7.1 to 7.0.0: https://github.com/atata-framework/atata-htmlvalidation/commit/612826009b667deb47b6ee38342c4b19b2f40311 So perhaps we need to adjust configs? I didn't see notes about that BTW: https://html-validate.org/changelog/index.html

sarahelsaig replied:
What changed in the config format? I only see new options and rules added in the change log you linked. That shouldn't be a problem.

Piedone replied:
Great then! @dministro what were you referring to?

Piedone replied:
@dministro?

Piedone commented on Lombiq.Tests.UI/Services/TestConfigurationManager.cs:
```diff
@@ -64,14 +64,15 @@ public static T GetConfiguration<T>()
+        // because AddJsonFile throws if the path is null or empty.
+        var sharedTestConfigurationPath =
+            Environment.GetEnvironmentVariable("LOMBIQ_UI_TESTING_TOOLBOX_SHARED_TEST_CONFIGURATION") ??
+            $"dummy_{Guid.NewGuid()}";
```
Instead of a dummy path that may cause issues later it would be better to just not call `AddJsonFile()` if not needed.

Piedone commented on Lombiq.Tests.UI/OrchardCoreUITestBase.cs:
```diff
@@ -262,17 +262,18 @@ public abstract class OrchardCoreUITestBase
         await UITestExecutor.ExecuteOrchardCoreTestAsync(testManifest, configuration);
     }
 
+    static OrchardCoreUITestBase() =>
+        ProgramCli.DefaultShellCliCommandFactory = OSDependentShellCliCommandFactory
+            .UseCmdForWindows()
+            .UseForOtherOS(new BashShellCliCommandFactory("-login"));
+
```
This would be better not here but in AtataFactory.

sarahelsaig replied:
We can't. Static constructors are called when the type is first referenced. So if we want to let the library consumer customize it in the `OrchardCoreUITestBase` overload as described in _Linux.md_, then we must be absolutely certain that this code has run before the setup customization. In general you should only use static constructor to set data that will be immediately interacted with. Every test calls to `OrchardCoreUITestBase` very early so that's perfect.
In this case I have verified that `AtataFactory`'s static constructor actually executes later than e.g. the configuration in `Lombiq.OSOCE.Tests.UI.UITestBase.ExecuteTestAsync`.

Piedone replied:
Then at least encapsulate this into `AtataFactory` so we don't have basic Atata configuration in multiple places.

Piedone committed:
Typo

sarahelsaig committed:
Instead of a dummy path that may cause issues later it would be better to just not call AddJsonFile() if not needed.

sarahelsaig committed:
Add SetupShellCliCommandFactory

sarahelsaig committed:
update package references

Piedone reviewed (changes requested):
Please add notes to the PR description about breaking changes, upgrade path, shared config option. This is to be added to the release notes later.

Piedone commented:
See my last comment.

sarahelsaig commented:
I have misread and added them to the issue description in Jira. Moved it here now.

dministro commented:
For Atata.HtmlValidation:
 - in Atata.HtmlValidation v1.4.0, the default version of html-validate(npm package) is [v5.2.0](https://github.com/atata-framework/atata-htmlvalidation/blob/43f1e948a247d4ce7a9a13782470a916ddc9be33/src/Atata.HtmlValidation/HtmlValidationOptions.cs#L120)
 - in Atata.HtmlValidation v2.0.0, the default version of html-validate(npm package) is [v7.0.0](https://github.com/atata-framework/atata-htmlvalidation/blob/2192e88f39cd2fcd163ce0b8f05ff74b86e58eaf/src/Atata.HtmlValidation/HtmlValidationOptions.cs#L123)

We not change the npm package version in [HtmlValidationConfiguration](https://github.com/Lombiq/UI-Testing-Toolbox/blob/cf9dc9ef8f25bb5bc9ea9197cd8d531fc470e18b/Lombiq.Tests.UI/Services/HtmlValidationConfiguration.cs#L29) but it would be good, to set it explicitly to avoid version issues in the future. So Atata.HtmlValidation uses its own default html-validate package version.

The html-validate configuration(.htmlvalidate.json) format was changed between the mentioned two versions, the [Lombiq.Tests.UI](https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI/.htmlvalidate.json), [Lombiq.Tests.UI.Samples](https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI.Samples/.htmlvalidate.json) and [Lombiq.OSOCE.NuGet.Tests.UI](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/blob/dev/NuGetTest/test/Lombiq.OSOCE.NuGet.Tests.UI/.htmlvalidate.json) are affected by this.

The html-validate configuration changes in [here](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/blob/a4c89e49f8619e515dfa6dcc6d051d43bc174b9c/NuGetTest/test/Lombiq.OSOCE.NuGet.Tests.UI/.htmlvalidate.json#L19-L28), [here](https://github.com/Lombiq/UI-Testing-Toolbox/blob/038a418eef44c9591c4d404e02f4a9f88c60c7e2/Lombiq.Tests.UI/.htmlvalidate.json#L18-L27) and [here](https://github.com/Lombiq/UI-Testing-Toolbox/blob/038a418eef44c9591c4d404e02f4a9f88c60c7e2/Lombiq.Tests.UI.Samples/.htmlvalidate.json#L20-L29) are ok for html-validate v5.2.0, but not for v7.0.0
For v7.0.0, the counfiguration should be something like below([based on this](https://gitlab.com/html-validate/html-validate/-/blob/v7.0.0/elements/html5.js#L528)):
```
{
...

  "elements": [
    "html5",
    {
      "form": {
        "attributes": {
          "method": {
			"enum": [ "get", "post", "GET", "POST" ]
		  }
        }
      }
    }
  ]
}
```

sarahelsaig committed:
Update html-validate config

based on https://github.com/Lombiq/UI-Testing-Toolbox/pull/119#issuecomment-1132134078

sarahelsaig committed:
Fix syntax.

sarahelsaig commented:
@dministro It doesn't seem to be the case. If I do it the way you suggested [I get a an error that says "A configuration error was found: TYPE must be array"](https://ci.lombiq.com/buildConfiguration/OrchardExtensions_OSOCE_Developer_PullBuildTestNuGetSolution/1029270?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true), but if I revert it to the way it is in your PR, then it works fine. I think the code you've linked from GitLab is a schema definition and not the literal structure of the _.htmlvalidate.json_ file.

Piedone commented on Lombiq.Tests.UI.Samples/.htmlvalidate.json:
```diff
@@ -15,5 +15,16 @@
     "wcag/h71": "off"
-  }
+  },
+
+  "elements": [
+    "html5",
+    {
+      "form": {
+        "attributes": {
+          "method": [ "get", "post", "GET", "POST" ]
+        }
+      }
+    }
+  ]
```
Do we actually need this config and Lombiq.Tests.UI/.htmlvalidate.json or did you add it for testing? Same for NuGetTest/test/Lombiq.OSOCE.NuGet.Tests.UI/.htmlvalidate.json.

sarahelsaig replied:
It will be needed for Jocó's issue so I see no reason to remove it, especially if his ends up merged sooner which is totally possible. If I were to delete these lines then git will get confused later during merge as one branch adds and another removes the same code. No thanks.

Piedone replied:
OK then.

sarahelsaig committed:
Use the new built-in features in the latest WebDriver.Net release.

sarahelsaig committed:
unusing

sarahelsaig committed:
Revert everything that doesn't need to be async

in AtataFactory and WebDriverFactory.

sarahelsaig committed:
unusing

sarahelsaig committed:
More neural language parser friendly comment in gitattributes.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

0liver commented on Lombiq.Tests.UI.Samples/Helpers/SetupHelpers.cs:
```diff
@@ -65,5 +68,27 @@ public static async Task<Uri> RunAutoSetupAsync(UITestContext context)
+
+            if (!validationErrors.Any()) throw;
+
+            var isSingle = validationErrors.Count == 1;
+            var plural = isSingle ? string.Empty : "s";
+            var errors = isSingle
+                ? " " + validationErrors.Single().Text.Trim()
+                : validationErrors
+                    .Select(element => "\n- " + element.Text.Trim())
+                    .Join(string.Empty);
+
+            throw new AssertionException($"Setup has failed with the following validation error{plural}:{errors}");
```
How slick 😁 And very nerdy, too 😉 

Would you despise having a newline and a hyphen when there's only a single error, or were you just having a good time 🥳? Because this could - as you most certainly know - become:

```suggestion
            var errors = "\n- " + validationErrors.Select(element => element.Text.Trim()).Join("\n- ");
            throw new AssertionException($"Setup has failed with the following validation errors:{errors}");
```

Or:

```suggestion
            var errors = new[] { string.Empty }.Concat(validationErrors).Select(element => element.Text.Trim()).Join("\n- ");
            throw new AssertionException($"Setup has failed with the following validation errors:{errors}");
```

sarahelsaig replied:
I didn't want to say "errors" when it only returns one error. Everything else is just flourish.

0liver commented on Lombiq.Tests.UI/Docs/Configuration.md:
````diff
@@ -73,27 +75,49 @@ If you have multiple UI test projects in a single solution and you're executing
+```
 
-You can use Docker Desktop to stop or start the container going forward. 
+If you haven't yet done it yet, add your user to the `docker` group.
````
Shorter, to the point, and correct:
```suggestion
If you haven't yet, add your user to the `docker` group.
```

0liver commented on Lombiq.Tests.UI/Docs/Configuration.md:
```diff
@@ -73,27 +75,49 @@ If you have multiple UI test projects in a single solution and you're executing
-You can use Docker Desktop to stop or start the container going forward. 
+If you haven't yet done it yet, add your user to the `docker` group.
+
+If you get `PlatformNotSupportedException`, that's a known problem with _Microsoft.Data.SqlClient_ on .Net 5 and above. As a workaround temporarily set the project's runtime identifier to `linux-x64` either [on the terminal](https://github.com/dotnet/SqlClient/issues/1423#issuecomment-1093430430) or by adding `<RuntimeIdentifier>linux-x64</RuntimeIdentifier>` to the project file.   
```
```suggestion
If you get a `PlatformNotSupportedException`, that's a known problem with _Microsoft.Data.SqlClient_ on .Net 5 and above. As a workaround, temporarily set the project's runtime identifier to `linux-x64` - either [on the terminal](https://github.com/dotnet/SqlClient/issues/1423#issuecomment-1093430430), or by adding `<RuntimeIdentifier>linux-x64</RuntimeIdentifier>` to the project file.   
```

0liver commented on Lombiq.Tests.UI/Docs/Configuration.md:
```diff
@@ -73,27 +75,49 @@ If you have multiple UI test projects in a single solution and you're executing
+
+#### On Both
+
+If you want to test it out, type `docker exec -u 0 -it sql2019 /opt/mssql-tools/bin/sqlcmd -S localhost -U sa -P "Password1!"` to access the SQL console.
```
Why can you use quotes on Linux here (around **Password1!**), but not in the docker installation scripts above?

sarahelsaig replied:
It should be apostrophe here too, because `!` has a special meaning under bash so we have to use apos to treat the string as a literal.

0liver commented on Lombiq.Tests.UI/Docs/Configuration.md:
```diff
@@ -73,27 +75,49 @@ If you have multiple UI test projects in a single solution and you're executing
+You can use [Docker Desktop](https://www.docker.com/products/docker-desktop) or [Portainer](https://www.portainer.io) to stop or start the container going forward. 
 
 
 ### Extending TestConfiguration.json
```
A single empty line suffices:
```suggestion

### Extending TestConfiguration.json
```

0liver commented on Lombiq.Tests.UI/Docs/Linux.md:
```diff
@@ -0,0 +1,25 @@
+
+As Linux has a stricter access policy you may want to install NPM in the userspace so you can still install global packages (e.g. html-validate) without sudoing. The easiest way to do this is via NVM. If you don't have NVM yet, [follow the guide here](https://github.com/Lombiq/NPM-Targets/tree/dev#global-npm-vs-userspace-npm-via-node-version-manager-on-linux).
+
+This library configures processes launched by Atata (via `Atata.Cli.ProgramCli`) to use Bash as login shell on non-Windows systems by default, like this
```
Colon(el), please!
```suggestion
This library configures processes launched by Atata (via `Atata.Cli.ProgramCli`) to use Bash as login shell on non-Windows systems by default, like this:
```

0liver commented on Lombiq.Tests.UI/Docs/Linux.md:
```diff
@@ -0,0 +1,25 @@
+
+## SQL Server Usage
+
+
```
Two empty lines before H2, and one after ([docs](https://lombiq.atlassian.net/wiki/spaces/DEV/pages/12156958/Adding+documentation+pages+to+our+projects#Linking-between-pages)):
```suggestion


## SQL Server Usage

```
(The preview is lying.)

0liver replied:
![image](https://user-images.githubusercontent.com/164360/170036189-badea9a6-d119-4c02-ba00-2ec6ee34fe94.png)

sarahelsaig committed:
Update html-validate version.

0liver commented on Lombiq.Tests.UI/Extensions/LoggingWebDriverExtensions.cs:
```diff
@@ -1,141 +1,38 @@
 namespace Lombiq.Tests.UI.Extensions;
 
-[DebuggerDisplay("{ToString()}")]
+[Obsolete($"Use {nameof(LogEntry)} instead.")]
```
How about adding a namespace to the suggested type name for clarity?

sarahelsaig replied:
I don't know of any way to get the namespace that keeps this variable `const` which is required. For example if I used `typeof(LogEntry).FullName` I'd get _[CS0182] An attribute argument must be a constant expression_ build error. And I don't want to hard code it becuase that's evil. Also you can just jump to declaration or mouseover.

0liver replied:
Like this?
```c#
[Obsolete($"Use {nameof(System)}{nameof(System.Text)}{nameof(Encoding)} instead.")]
```
(The example uses a different type, but you get the point. Is this pretty? No. Does it help when you read the warning message? I suppose.)

sarahelsaig replied:
Don't like it. This is just hard coding with extra steps. Also you forgot the dots between the names.

0liver commented on Lombiq.Tests.UI/Extensions/StringExtensions.cs:
```diff
@@ -1,12 +1,62 @@
+    }
+        .Concat(Path.GetInvalidFileNameChars())
+        .Concat(Path.GetInvalidPathChars())
+        .Distinct()
```
`.Union()`?

0liver replied:
And why are the character sets in `Path.GetInvalidFileNameChars()` and `Path.GetInvalidPathChars()` not sufficient in the first place?

sarahelsaig replied:
Because they return different values on different OSes. Specifically, `actions/upload-artifact@v2` throws an error because `:` is not stripped from the failure dump path on Linux if you only use `Path.GetInvalidFileNameChars()` and `Path.GetInvalidPathChars()`. It's actually a valid path character (at least on ext4) so that's only a problem when you try to copy the resulting directory to an NTFS drive. 

Strange, I was sure I have added a comment explaining it, maybe it didn't get saved. I wrote a new one:

> The first array contains every invalid character in Windows. This is the union of GetInvalidFileNameChars and
GetInvalidPathChars on Windows (for FAT and NTFS). Linux file systems aren't as strict, they only forbid forward
slash ('/') and null character ('\0'). This is a subset of the invalid characters on Windows. So to make the
paths work across all systems (e.g. when we plan to back up to an NTFS store or use GitHub's
"actions/upload-artifact" action) we have to carry the banned Windows characters across operating systems.
GetInvalidFileNameChars and GetInvalidPathChars are still included in case they contain more characters on
untested operating systems.

0liver replied:
I remember that discussion now. Thanks again for the insight - the comment lays it out quite nicely.

0liver commented on Lombiq.Tests.UI/Extensions/StringExtensions.cs:
```diff
@@ -1,12 +1,62 @@
 
 public static class StringExtensions
 {
+    private static readonly Lazy<char[]> _invalidPathCharacters = new(() =>
```
A lazily instantiated array - neat!

0liver commented on Lombiq.Tests.UI/Extensions/VerificationUITestContextExtensions.cs:
```diff
@@ -9,4 +11,39 @@ public static class VerificationUITestContextExtensions
     public static PageNavigationState AsPageNavigationState(this UITestContext context) => new(context);
+
+    /// <summary>
+    /// Ensures all the logs pass successfully and throws if they didn't.
```
I don't exactly understand this part:
> the logs pass successfully

What is it supposed to mean?

sarahelsaig replied:
reworded: Verifies all logs and throws an exception if they didn't pass the checks.

0liver commented on Lombiq.Tests.UI/Services/AtataFactory.cs:
```diff
@@ -24,62 +25,82 @@ public static class AtataFactory
+        TimeoutConfiguration timeoutConfiguration,
+        ITestOutputHelper testOutputHelper)
+    {
+        IWebDriver CastDriverFactory<T>(Func<BrowserConfiguration, TimeSpan, T> factory)
```
The purpose of this method is not really to cast - it is to invoke the provided driver generator. I suggest to simply call it `Invoke` which will lead to the usages [below](https://github.com/Lombiq/UI-Testing-Toolbox/pull/119/files#diff-5128e49048e608f35873020d110b968da3c1dddd81746b83674c4209451b0335R74) looking like this:
```c#
Browser.Chrome => Invoke(WebDriverFactory.CreateChromeDriver),
```

`From` seems like another valid option - do you have a better one? In any case, please rename this method to reflect its purpose.

sarahelsaig replied:
`From` sounds cool

0liver commented on Lombiq.Tests.UI/Services/AtataFactory.cs:
```diff
@@ -24,62 +25,82 @@ public static class AtataFactory
+                    currentTryIndex++;
+                    var retryCount = maxTryCount - currentTryIndex;
+
+                    // Not using parameters because the exception can throw off the string format.
```
How exactly?

sarahelsaig replied:
Beats me, [it's not my code](https://github.com/Lombiq/UI-Testing-Toolbox/commit/cd3123fcce88b2cc4911d57568a60d54fd3989b0)

@Piedone , the inquisition is here for you :D

0liver commented on Lombiq.Tests.UI/Services/AtataFactory.cs:
```diff
@@ -24,62 +25,82 @@ public static class AtataFactory
+        }
+    }
+
+    public static void SetupShellCliCommandFactory() =>
```
Apply the file layout refactoring again - this method should come before the `private static` one above.

sarahelsaig replied:
I have a memory that private static methods are supposed to come before public static methods so that was intentional. [The dojo](https://orcharddojo.net/orchard-resources/CoreLibrary/DevelopmentGuidelines/CodeStyling) is a bit ambiguous on the topic now that I look at it. @Piedone what is your opinion?

0liver replied:
I've interpolated the rule from two rules that we do have in the Dojo styleguide:

1. Public instance methods come before private instance methods.
2. Static methods come after instance methods.

I think the point is that all private field and properties come at the beginning of the file, and all private methods at the end.

sarahelsaig replied:
Okay, I have swapped them for now. But Dojo should be clearer so we don't have to interpolate rules.

Piedone replied:
Yeah, public static methods should come before private static ones. Please submit a PR for it: https://github.com/Lombiq/Orchard-Dojo-Library

0liver commented on Lombiq.Tests.UI/Services/AtataFactory.cs:
```diff
@@ -24,62 +25,82 @@ public static class AtataFactory
+            catch (WebDriverException ex)
+            {
+                if (ex.Message.ContainsOrdinalIgnoreCase("Cannot start the driver service on") &&
+                    currentTryIndex < maxTryCount - 1)
```
In a case like this I'd prefer to use a 1-based counter.

Or rename the variables to `currentRetry` and `maxRetryCount = 2` and drop the `- 1`.

sarahelsaig replied:
changed it to `currentTry` using the 1 based index

0liver replied:
The code reads nicer now.

0liver commented on Lombiq.Tests.UI/Services/OrchardCoreInstance.cs:
```diff
@@ -104,14 +109,14 @@ public async Task<Uri> StartUpAsync()
         }
 
-        // If you try to use dotnet.exe to run a DLL published for a different platform (seen this with running
+        // If you try to use the dotnet command to run a DLL published for a different platform (seen this with running
         // win10-x86 DLLs on an x64 Windows machine) then you'll get a "Failed to load the dll from hostpolicy.dll
         // HRESULT: 0x800700C1" error even if the exe will run without issues. So, if an exe exists, we'll run that.
```
Let's be consistent with DLL for improved readability:
```suggestion
        // If you try to use the dotnet command to run a DLL published for a different platform (seen this with running
        // win10-x86 DLLs on an x64 Windows machine) then you'll get a "Failed to load the dll from hostpolicy.dll
        // HRESULT: 0x800700C1" error even if the EXE will run without issues. So, if an EXE exists, we'll run that.
```

sarahelsaig replied:
But not too consistent.
![image](https://user-images.githubusercontent.com/4007293/170149292-e769a223-5e02-4568-a5f1-14a0a4e415ec.png)

I think we never capitalize extensions anywhere else so probably that "DLLs" is what should be replaced.

0liver replied:
I didn't replace that one because it's inside an error message 😁

0liver replied:
I find dll and exe hard to read in the middle of a sentence to be honest.

sarahelsaig replied:
Agreed, but we shouldn't break convention just here.

0liver replied:
Yeah, that's the hardest part. The go ahead and make it consistent in some way, at least.

Commit:
Formatting, minor improvements

0liver commented on Lombiq.Tests.UI/Services/OrchardCoreInstance.cs:
```diff
@@ -139,9 +144,7 @@ public async Task<Uri> StartUpAsync()
                 });
         }
 
-        var argumentsBuilder = new ArgumentsBuilder();
-
-        argumentsBuilder = argumentsBuilder
```
I've contracted this - please double-check.

0liver commented on Lombiq.Tests.UI/Services/OrchardCoreInstance.cs:
```diff
@@ -191,16 +195,16 @@ public IEnumerable<IApplicationLog> GetLogs(CancellationToken cancellationToken
-            Directory
+        return Directory.Exists(logFolderPath)
+            ? Directory
                 .EnumerateFiles(logFolderPath, "*.log")
                 .Select(filePath => (IApplicationLog)new ApplicationLog
                 {
                     Name = Path.GetFileName(filePath),
                     FullName = Path.GetFullPath(filePath),
                     ContentLoader = () => File.ReadAllTextAsync(filePath, cancellationToken),
-                }) :
-            Enumerable.Empty<IApplicationLog>();
+                })
+            : Enumerable.Empty<IApplicationLog>();
```
Reformatted this, too.

0liver commented on Lombiq.Tests.UI/Services/OrchardCoreInstance.cs:
```diff
@@ -270,9 +278,9 @@ private Task StopOrchardAppAsync()
-        public string Name { get; set; }
-        public string FullName { get; set; }
-        public Func<Task<string>> ContentLoader { get; set; }
+        public string Name { get; init; }
+        public string FullName { get; init; }
+        public Func<Task<string>> ContentLoader { get; init; }
```
R# makes me (look) smart 😁

sarahelsaig replied:
![image](https://user-images.githubusercontent.com/4007293/170211396-ef856f27-2eed-40aa-8e14-932774e0501b.png)

0liver commented on Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs:
```diff
@@ -179,15 +180,16 @@ public void AssertBrowserLogMaybe(IList<BrowserLogMessage> browserLogs, Action<s
-    public static readonly Func<BrowserLogMessage, bool> IsValidBrowserLogMessage =
+    public static readonly Func<LogEntry, bool> IsValidBrowserLogMessage =
         message =>
+            message.Level >= LogLevel.Warning &&
```
Why are only warnings valid browser log messages?

sarahelsaig replied:
warnings and errors. logs and infos are not relevant to the UI Testing Toolbox

sarahelsaig committed:
Remove no longer needed configuration.

sarahelsaig committed:
Merge branch 'issue/OSOE-49' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-49

0liver commented on Lombiq.Tests.UI/Services/UITestExecutionSession.cs:
```diff
@@ -305,13 +303,14 @@ private async Task CaptureAppSnapshotAsync(string dumpContainerPath)
+                    remotePath,
+                    appDumpPath,
+                    containerName,
+                    useCompressionIfAvailable: true);
```
How about a more readable version:

```suggestion
                var containerName = _dockerConfiguration?.ContainerName;
                var remotePath = string.IsNullOrEmpty(containerName)
                    ? appDumpPath
                    : _dockerConfiguration.ContainerSnapshotPath;

                await _sqlServerManager.TakeSnapshotAsync(
                    remotePath,
                    appDumpPath,
                    containerName,
                    useCompressionIfAvailable: true);
```
**Remark**: I'm just referring to the first 4 lines in comparison to your first 3 - the rest is me fighting GitHub.

0liver replied:
Just found [this occurrence](https://github.com/Lombiq/UI-Testing-Toolbox/pull/119/files#diff-7296bb52a59df5fa30fd8222aa55a27fb459709e7e0cb6539be92a6ac2606069R601) where you did exactly the same thing 😀

Commit:
Code formatting, comment improvements

Commit:
Fix class member order

0liver commented on Lombiq.Tests.UI/Services/UITestExecutionSession.cs:
```diff
@@ -23,20 +24,16 @@ namespace Lombiq.Tests.UI.Services;
 
 internal sealed class UITestExecutionSession : IAsyncDisposable
 {
+    // We need to have different snapshots based on whether the test uses the defaults, SQL Server and/or Azure Blob.
+    private static readonly ConcurrentDictionary<string, SynchronizingWebApplicationSnapshotManager> _setupSnapshotManagers = new();
+    private static readonly ConcurrentDictionary<string, (int FailureCount, Exception LatestException)> _setupOperationFailureCount = new();
```
I've moved this here where it belongs.

0liver commented on Lombiq.Tests.UI/Services/UITestExecutionSession.cs:
```diff
@@ -432,70 +430,55 @@ private async Task SetupAsync()
+        // Thus, if anything goes wrong between the below delegate registration and its invocation, it will remain
+        // registered and fail on the disposed SqlServerManager during a retry.
         _configuration.OrchardCoreConfiguration.BeforeTakeSnapshot =
-            _configuration.OrchardCoreConfiguration.BeforeTakeSnapshot.RemoveAll(
-                SqlServerManagerBeforeTakeSnapshotHandlerAsync);
-        _configuration.OrchardCoreConfiguration.BeforeTakeSnapshot +=
-            SqlServerManagerBeforeTakeSnapshotHandlerAsync;
+            _configuration.OrchardCoreConfiguration.BeforeTakeSnapshot.RemoveAll(SqlServerManagerBeforeTakeSnapshotHandlerAsync);
+        _configuration.OrchardCoreConfiguration.BeforeTakeSnapshot += SqlServerManagerBeforeTakeSnapshotHandlerAsync;
```
This was me, too.

0liver commented on Lombiq.Tests.UI/Services/UITestExecutionSession.cs:
```diff
@@ -573,10 +556,7 @@ Task UITestingBeforeAppStartHandlerAsync(string contentRootPath, ArgumentsBuilde
-            _testOutputHelper,
-            uri,
-            _configuration);
+        var atataScope = AtataFactory.StartAtataScope(_testOutputHelper, uri, _configuration);
```
And this.

0liver commented on Lombiq.Tests.UI/Services/UITestExecutionSession.cs:
```diff
@@ -632,16 +583,17 @@ async Task SqlServerManagerBeforeAppStartHandlerAsync(string contentRootPath, Ar
-                _snapshotDirectoryPath;
-
-            if (!_hasSetupOperation || !Directory.Exists(_dockerConfiguration?.HostSnapshotPath ?? snapshotDirectoryPath))
+            if (!_hasSetupOperation || !Directory.Exists(_snapshotDirectoryPath))
```
I've removed the local variable `snapshotDirectoryPath` completely, just like in the method below.

0liver commented on Lombiq.Tests.UI/Services/WebDriverFactory.cs:
```diff
@@ -18,28 +18,47 @@ public static class WebDriverFactory
 
-                if (configuration.Headless) options.AddArgument("headless");
+            // Disabling the Chrome sandbox can speed things up a bit, so recommended when you get a lot of timeouts
+            // during parallel execution:
```
```suggestion
            // Disabling the Chrome sandbox can speed things up a bit, so it's recommended when you get a lot of
            // timeouts during parallel execution:
```

0liver commented on Lombiq.Tests.UI/Services/WebDriverFactory.cs:
```diff
@@ -18,28 +18,47 @@ public static class WebDriverFactory
 
-                configuration.BrowserOptionsConfigurator?.Invoke(options);
+            // Linux-specific setting, may be necessary for running in containers, see
+            // https://developers.google.com/web/tools/puppeteer/troubleshooting#tips
```
```suggestion
           // https://developers.google.com/web/tools/puppeteer/troubleshooting#tips for more information.
```
😉

0liver commented on Lombiq.Tests.UI/Services/WebDriverFactory.cs:
```diff
@@ -18,28 +18,47 @@ public static class WebDriverFactory
+        }
+
+        if (Environment.GetEnvironmentVariable("CHROMEWEBDRIVER") is { } driverPath &&
+            Directory.Exists(driverPath))
```
Would you mind putting this on a single line?

0liver commented on Lombiq.Tests.UI/Services/WebDriverFactory.cs:
```diff
@@ -18,28 +18,47 @@ public static class WebDriverFactory
-                var options = new ChromeOptions().SetCommonOptions();
+    public static ChromeDriver CreateChromeDriver(BrowserConfiguration configuration, TimeSpan pageLoadTimeout)
+    {
+        var state = new ChromeConfiguration { Options = new ChromeOptions().SetCommonOptions(), Service = null };
```
`config` looks better to me here than `state`.

0liver replied:
Also, you don't need to set `Service = null` here.

Commit:
Refactoring CreateDriverInner to avoid external state

0liver commented on Lombiq.Tests.UI/Services/WebDriverFactory.cs:
```diff
@@ -18,28 +18,47 @@ public static class WebDriverFactory
+        if (Environment.GetEnvironmentVariable("CHROMEWEBDRIVER") is { } driverPath &&
+            Directory.Exists(driverPath))
+        {
+            state.Service = ChromeDriverService.CreateDefaultService(driverPath);
+            return CreateDriverInner();
```
I would prefer to move the `config` (`state`) variable into `CreateDriverInner()` and invoke the latter with a new parameter `service` which will be set to `ChromeDriverService.CreateDefaultService(driverPath)` here, and ... see my PR: https://github.com/Lombiq/UI-Testing-Toolbox/pull/167.

Commit:
Better readability thanks to switching parameters around

0liver commented on Lombiq.Tests.UI/Services/WebDriverFactory.cs:
```diff
@@ -129,36 +141,39 @@ private static TDriver SetCommonTimeouts<TDriver>(this TDriver driver, TimeSpan
     }
 
     private static TDriver CreateDriver<TDriver>(Func<TDriver> driverFactory, IDriverConfig driverConfig)
```
I've allowed myself for some refactoring of this methods which leads to more readable code. See my PR: https://github.com/Lombiq/UI-Testing-Toolbox/pull/167. All UI tests passed successfully locally.

sarahelsaig committed:
Rename state to config.

sarahelsaig committed:
Merge branch 'issue/OSOE-49' into issue/OSOE-49-suggestion

# Conflicts:
#	Lombiq.Tests.UI/Services/WebDriverFactory.cs

sarahelsaig committed:
Update Lombiq.Tests.UI/Docs/Configuration.md

Co-authored-by: 0liver <0liver@users.noreply.github.com>

sarahelsaig committed:
Update Lombiq.Tests.UI/Docs/Configuration.md

Co-authored-by: 0liver <0liver@users.noreply.github.com>

sarahelsaig committed:
Replace quotes with apostrophes

Commit:
Rename config variable

sarahelsaig committed:
Apply suggestions from code review

Co-authored-by: 0liver <0liver@users.noreply.github.com>

sarahelsaig committed:
Rename config to chromeConfig.

sarahelsaig committed:
Merge branch 'issue/OSOE-49' into issue/OSOE-49-suggestion

sarahelsaig committed:
Merge pull request #167 from Lombiq/issue/OSOE-49-suggestion

OSOE-49: Refactoring suggestion

sarahelsaig committed:
StringExtensions comment and union.

sarahelsaig committed:
Reword doc.

sarahelsaig committed:
Rename CastDriverFactory to From

sarahelsaig committed:
Rename to currentTry

sarahelsaig commented:
I have "resolved" the comments regarding the changes you made. I think it would've made more sense to include those as comments in your own PR.

sarahelsaig committed:
Make the error list uglier on purpose.

Co-authored-by: 0liver <0liver@users.noreply.github.com>

sarahelsaig committed:
Code cleanup.

sarahelsaig committed:
Merge branch 'issue/OSOE-49' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-49

sarahelsaig committed:
fix error

sarahelsaig committed:
Change method order.

0liver commented on Lombiq.Tests.UI/Extensions/StringExtensions.cs:
```diff
@@ -1,12 +1,68 @@
 
 public static class StringExtensions
 {
+    // The first array contains every invalid character in Windows. This is the union of GetInvalidFileNameChars and
+    // GetInvalidPathChars on Windows (for FAT and NTFS). Linux file systems aren't as strict, they only forbid forward
+    // slash ('/') and null character ('\0'). This is a subset of the invalid characters on Windows. So to make the
+    // paths work across all systems (e.g. when we plan to back up to an NTFS store or use GitHub's
+    // "actions/upload-artifact" action) we have to carry the banned Windows characters across operating systems.
+    // GetInvalidFileNameChars and GetInvalidPathChars are still included in case they contain more characters on
+    // untested operating systems.
```
I've improved your excellent comment by just some jotas:
```suggestion
    // The first array contains every invalid character in Windows. This is the union of GetInvalidFileNameChars() and
    // GetInvalidPathChars() evaluated on Windows (for FAT and NTFS). Linux file systems aren't as strict, they only
    // forbid forward slash ('/') and null character ('\0'). This is a subset of the invalid characters on Windows. So
    // to make the paths work across all systems (e.g. when we plan to back up to an NTFS store or use GitHub's
    // "actions/upload-artifact" action) we have to carry the banned Windows characters across operating systems.
    // GetInvalidFileNameChars() and GetInvalidPathChars() are still included in case they contain more characters on
    // untested operating systems.
```

0liver commented on Lombiq.Tests.UI/Extensions/StringExtensions.cs:
```diff
@@ -1,12 +1,68 @@
+        '\0',
+        '\u0001',
+        '\u0002',
+        '\u0003',
+        '\u0004',
+        '\u0005',
+        '\u0006',
+        '\a',
+        '\b',
+        '\t',
+        '\n',
+        '\v',
+        '\f',
+        '\r',
+        '\u000e',
+        '\u000f',
+        '\u0010',
+        '\u0011',
+        '\u0012',
+        '\u0013',
+        '\u0014',
+        '\u0015',
+        '\u0016',
+        '\u0017',
+        '\u0018',
+        '\u0019',
+        '\u001a',
+        '\u001b',
+        '\u001c',
+        '\u001d',
+        '\u001e',
+        '\u001f',
+        '|',
+        ':',
+        '"',
+        '<',
+        '>',
+        '*',
+        '?',
+    }
```
The formatting didn't seem quite right, so I've reformatted it as follows, so that the chained method calls align more prettily:
```suggestion
    private static readonly Lazy<char[]> _invalidPathCharacters = new(() => new[]
        {
            '\0',
            '\u0001',
            '\u0002',
            '\u0003',
            '\u0004',
            '\u0005',
            '\u0006',
            '\a',
            '\b',
            '\t',
            '\n',
            '\v',
            '\f',
            '\r',
            '\u000e',
            '\u000f',
            '\u0010',
            '\u0011',
            '\u0012',
            '\u0013',
            '\u0014',
            '\u0015',
            '\u0016',
            '\u0017',
            '\u0018',
            '\u0019',
            '\u001a',
            '\u001b',
            '\u001c',
            '\u001d',
            '\u001e',
            '\u001f',
            '|',
            ':',
            '"',
            '<',
            '>',
            '*',
            '?',
        }
```

0liver commented on Lombiq.Tests.UI/Services/AtataFactory.cs:
```diff
@@ -24,62 +25,79 @@ public static class AtataFactory
+                }
+
+                currentTry++;
+                var retryCount = maxTryCount - currentTry + 1;
```
On the first try, we should have 2 more tries, which would be:
```suggestion
                var retryCount = maxTryCount - currentTry;
```
Just like before.

0liver commented on Lombiq.Tests.UI/Services/OrchardCoreInstance.cs:
```diff
@@ -53,6 +55,7 @@ public sealed class OrchardCoreInstance : IWebApplicationInstance
     private static readonly PortLeaseManager _portLeaseManager;
     private static readonly ConcurrentDictionary<string, string> _exeCopyMarkers = new();
     private static readonly object _exeCopyLock = new();
+    private static readonly bool _isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
```
Now, I've noticed it: We use `OperatingSystem.IsWindows()` in `Lombiq.HelpfulLibraries.Cli.Helpers.OperatingSystemHelper.GetExecutableExtension()`, but here we have this. Any specific reason for this choice?

~Googling~ Ducking this led me to the following interesting comment under [this answer](https://stackoverflow.com/a/38790956/177710):
![image](https://user-images.githubusercontent.com/164360/170370765-631e9bdb-bfef-4bef-a233-171a2dda8d68.png)

Does it have to do with that?

sarahelsaig replied:
I think the term you are looking for is... "searching"? 😁 

I can't think of a particular reason, and `OperatingSystem.IsWindows()` uses the build target to determine the OS as well:
![image](https://user-images.githubusercontent.com/4007293/170374717-d9b27752-dd7c-4013-bf8d-7dc420fa8120.png)

And IsOSPlatform:
![image](https://user-images.githubusercontent.com/4007293/170375090-929495c8-49d7-4f30-801c-0a59ede4a8d0.png)

Since both use preprocessor directives there is no functionality difference, only an imperceptibly small difference as `IsOSPlatform` does its additional string check. Not even worth thinking about.

I've changed `_isWindows ` to use `OperatingSystem.IsWindows()` for simplicity.

0liver replied:
I have to become bolder at simply looking at the source, too!

0liver commented on Lombiq.Tests.UI/Services/OrchardCoreInstance.cs:
```diff
@@ -63,6 +66,8 @@ public sealed class OrchardCoreInstance : IWebApplicationInstance
     private string _contentRootPath;
     private bool _isDisposed;
 
+    private static string ExecutableExtension => _isWindows ? ".exe" : string.Empty;
```
Seeing it now, this could actually be just another static field instead of a property, right underneath `_isWindows`.

0liver replied:
You could even inline `_isWindows` for that matter.

sarahelsaig replied:
Originally it was a property, because one field can't be initialized from another field's value. I like inlining `_isWindows` though. I think I had some other uses for `_isWindows` at some point but not any more.

0liver replied:
Actually, you can if only you initialize them in the correct order:

![image](https://user-images.githubusercontent.com/164360/170380890-6fc1efae-be7f-48ee-a98e-124596841134.png)

sarahelsaig committed:
Update Lombiq.Tests.UI/Extensions/StringExtensions.cs

Co-authored-by: 0liver <0liver@users.noreply.github.com>

sarahelsaig committed:
Update Lombiq.Tests.UI/Extensions/StringExtensions.cs

Co-authored-by: 0liver <0liver@users.noreply.github.com>

sarahelsaig committed:
Update Lombiq.Tests.UI/Services/AtataFactory.cs

Co-authored-by: 0liver <0liver@users.noreply.github.com>

sarahelsaig committed:
Simplify OrchardCoreConfiguration._isWindows initialization.

sarahelsaig committed:
Simplify ExecutableExtension.

sarahelsaig committed:
Fix R# "possible null assignment" warnings in OrchardCoreConfiguration.
